llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -1015,6 +1017,81 @@ void SpirvLowerGlobal::lowerInOutInPlace() {
   }
 }
 
+// =====================================================================================================================
+// @param builtIn : BuiltIn value
+// @param elemIdx : Element Index of struct
+Value *SpirvLowerGlobal::createRaytracingBuiltIn(BuiltIn builtIn, Value *elemIdx) {
+  Value *builtinValue = nullptr;
+  switch (builtIn) {
+  case BuiltInLaunchIdKHR:
+    builtinValue = m_builder->create<DispatchRaysIndexOp>();
+    break;
+  case BuiltInLaunchSizeKHR:
+    builtinValue = m_builder->create<DispatchRaysDimensionsOp>();
+    break;
+  case BuiltInWorldRayOriginKHR:
+    builtinValue = m_builder->create<WorldRayOriginOp>();
+    break;
+  case BuiltInWorldRayDirectionKHR:
+    builtinValue = m_builder->create<WorldRayDirectionOp>();
+    break;
+  case BuiltInObjectRayOriginKHR:
+    builtinValue = m_builder->create<ObjectRayOriginOp>();
+    break;
+  case BuiltInObjectRayDirectionKHR:
+    builtinValue = m_builder->create<ObjectRayDirectionOp>();
+    break;
+  case BuiltInRayTminKHR:
+    builtinValue = m_builder->create<RayTminOp>();
+    break;
+  case BuiltInRayTmaxKHR:
+    builtinValue = m_builder->create<RayTcurrentOp>();
+    break;
+  case BuiltInInstanceCustomIndexKHR:
+    builtinValue = m_builder->create<InstanceIndexOp>();
+    break;
+  case BuiltInObjectToWorldKHR:
+    builtinValue = m_builder->create<ObjectToWorldOp>();
+    break;
+  case BuiltInWorldToObjectKHR:
+    builtinValue = m_builder->create<WorldToObjectOp>();
+    break;
+  case BuiltInHitKindKHR:
+    builtinValue = m_builder->create<HitKindOp>();
+    break;
+  case BuiltInHitTriangleVertexPositionsKHR:
+    builtinValue = m_builder->create<TriangleVertexPositionsOp>(elemIdx);
+    break;
+  case BuiltInIncomingRayFlagsKHR:
+    builtinValue = m_builder->create<RayFlagsOp>();
+    break;
+  case BuiltInRayGeometryIndexKHR:
+    builtinValue = m_builder->create<GeometryIndexOp>();
+    break;
+  case BuiltInInstanceId:
+    builtinValue = m_builder->create<InstanceIdOp>();
+    break;
+  case BuiltInPrimitiveId:
+    builtinValue = m_builder->create<PrimitiveIndexOp>();
+    break;
+  }
+
+  assert(builtinValue != nullptr);
```

Comment:
When can this fail? It seems that calling this function relies on the ordering of values in the enum containing `BuiltInLaunchIdKHR` and `BuiltInRayGeometryIndexKHR`? What happens if this condition no longer applies because values in the enum are re-ordered? Will this return a `nullptr` in release mode that is not handled after the call to `createRaytracingBuiltIn`?

Reply:
>>When can this fail? 
        when there is unhandled builtin passed in.
>> What happens if this condition no longer applies because values in the enum are re-ordered?
     For spirv, if there is assigned enum,  their values does not change.
>> Will this return a nullptr in release mode that is not handled after the call to createRaytracingBuiltIn?
     yes, in this case , we will let the compiler crash and fail